django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,130 @@
+from django.db.models import Q
+
+
+def user_can_edit_chants_in_source(user, source) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_is_proofreader(user) -> bool:
+    """
+    Checks if the user is a proofreader.
+    Used in UserSourceListView.
+    """
+    return user.groups.filter(Q(name="project manager") | Q(name="editor")).exists()
+
+
+def user_can_proofread_chants_in_source(user, source_id) -> bool:
+    """
+    Checks if the user can access the proofreading page of a given Source.
+    Used in ChantProofreadView.
+    """
+    if user.is_anonymous:
+        return False
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_id_editor: bool = user.groups.filter(name="editor").exists()
+
+    if (user_is_project_manager) or (user_id_editor and user_is_assigned_to_source):
```

Comment:
`return (user_is_project_manager) or (user_id_editor and user_is_assigned_to_source)`

Reply:
this can be used in all places where you currently end a function with
```python
if condition:
    return True
return False
```
With `return condition`, you can accomplish the same thing with the same clarity on one line rather than three. I'll stop commenting on this whenever I come across it.